### PR TITLE
Feature/search parameter units

### PR DESCRIPTION
**Pull request by timosachsenberg**
addresses https://github.com/OpenMS/OpenMS/issues/1435

### CODE REVIEW

**Comment by timosachsenberg** [comment]
Currently we don't write the unit information to our data formats but in mzIdentML and mzTab it is intended.

**Comment by timosachsenberg** [comment]
travis passed - then one cppcheck error was from other PR

File: src/topp/XTandemAdapter.cpp [hunk 1]
```diff
@@ -432,7 +432,7 @@ class TOPPXTandemAdapter :
     search_parameters.fixed_modifications = getStringList_("fixed_modifications");
     search_parameters.variable_modifications = getStringList_("variable_modifications");
     search_parameters.missed_cleavages = getIntOption_("missed_cleavages");
-    search_parameters.peak_mass_tolerance = getDoubleOption_("fragment_mass_tolerance");
+    search_parameters.fragment_mass_tolerance = getDoubleOption_("fragment_mass_tolerance");
```
Comment (cbielow) [inline src/topp/XTandemAdapter.cpp]: shouldnt we also set
```
bool fragment_mass_tolerance_ppm;
bool precursor_mass_tolerance_ppm;
```
using the XTandemAdapter parameters now that its available?

other issue:
"precursor_mass_tolerance" in the adapter is currently "1.5", but in ppm. This seems a rather stringent error windows and was probably ok for Da earlier. Could you increase that to ~20ppm?!

File: src/topp/OMSSAAdapter.cpp [hunk 2]
```diff
@@ -1005,9 +1005,11 @@ class TOPPOMSSAAdapter :
     search_parameters.variable_modifications = getStringList_("variable_modifications");
     search_parameters.digestion_enzyme = *EnzymesDB::getInstance()->getEnzyme(getStringOption_("enzyme"));
     search_parameters.missed_cleavages = getIntOption_("v");
-    search_parameters.peak_mass_tolerance = getDoubleOption_("fragment_mass_tolerance");
+    search_parameters.fragment_mass_tolerance = getDoubleOption_("fragment_mass_tolerance");
     search_parameters.precursor_tolerance = getDoubleOption_("precursor_mass_tolerance");
```
Comment (cbielow) [inline src/topp/OMSSAAdapter.cpp]: same here (set units)

File: src/topp/InspectAdapter.cpp [hunk 3]
```diff
@@ -848,7 +848,7 @@ class TOPPInspectAdapter :
           if (inspect_infile.getEnzyme() == "Trypsin") sp.enzyme = ProteinIdentification::TRYPSIN;
           else if (inspect_infile.getEnzyme() == "No_Enzyme") sp.enzyme = ProteinIdentification::NO_ENZYME;
           else sp.enzyme = ProteinIdentification::UNKNOWN_ENZYME;
-          sp.peak_mass_tolerance = inspect_infile.getPeakMassTolerance();
+          sp.fragment_mass_tolerance = inspect_infile.getPeakMassTolerance();
           sp.precursor_tolerance = inspect_infile.getPrecursorMassTolerance();
```
Comment (cbielow) [inline src/topp/InspectAdapter.cpp]: everywhere... basically :)

File: src/openms/source/FORMAT/HANDLERS/MascotXMLHandler.cpp [hunk 4]
```diff
@@ -545,7 +545,15 @@ namespace OpenMS
...
       else if (tag_ == "ITOL")
       {
-        search_parameters_.peak_mass_tolerance = (character_buffer_.trim()).toDouble();
+        search_parameters_.fragment_mass_tolerance = (character_buffer_.trim()).toDouble();
+      }
+      else if (tag_ == "TOLU")
+      {
+        search_parameters_.precursor_mass_tolerance_ppm = (character_buffer_.trim()) == "ppm";
+      }
+      else if (tag_ == "ITOLU")
```
Comment (cbielow) [inline src/openms/source/FORMAT/HANDLERS/MascotXMLHandler.cpp]: is it possible to extend the tests for this?

**Comment by cbielow** [comment]
I just realized that we currently to not store the unit when writing SearchParam to a file.
It seems a lot of work to fix this.

**Comment by timosachsenberg** [comment]
Jep. But I don't plan to fix it for idxml etc.

File: src/openms/source/FORMAT/HANDLERS/MascotXMLHandler.cpp (hunk 4 above)
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/MascotXMLHandler.cpp]: will check

File: src/topp/OMSSAAdapter.cpp (hunk 2 above)
Comment (timosachsenberg) [inline src/topp/OMSSAAdapter.cpp]: isn't it there (see line below)

Comment (timosachsenberg) [inline src/topp/OMSSAAdapter.cpp]: see line 1010, 1011 or what do you mean?

File: src/topp/XTandemAdapter.cpp (hunk 1 above)
Comment (timosachsenberg) [inline src/topp/XTandemAdapter.cpp]: ok I misse that one... thanks for pointing it out

**Comment by timosachsenberg** [comment]
TODO:
- set unit in XTandemAdapter 
- check MyriMatch, MSGF+
- adjust default config to something reasonable (10 ppm precursor tolerance, 0.3 Da fragment tolerance (or should be assume high-res HCD?))

**Comment by hendrikweisser** [comment]
> Jep. But I don't plan to fix it for idxml etc. 

So what's the plan for idXML? Store just the numeric value of the tolerance, or first convert it to Da if it's ppm?

And in the other direction, when reading idXML: Always use Da as the unit (then we should also convert to Da when writing), or add a heuristic, e.g. "Da if < 1, else ppm"?

**Comment by timosachsenberg** [comment]
There is no way we can do this right I guess. As current tools don't use this information except on export/conversion to mzIdentML and mzTab I guess we keep this as a preparative change. Proper use (e.g. to set parameters in the annotation view etc.) makes most sense when we got the transition to mzIdentML.
In mzTab we have the choice to just not report this information (yet) as it is optional. In mzIdentML I don't know - @mwalzer do you know by heart if this is mandatory?

**Comment by mwalzer** [comment]
<cvParam accession="MS:1001412" cvRef="PSI-MS" unitCvRef="UO" unitName="parts per million" unitAccession="UO:0000169" value="20.0" name="search tolerance plus value"/>
            <cvParam accession="MS:1001413" cvRef="PSI-MS" unitCvRef="UO" unitName="parts per million" unitAccession="UO:0000169" value="20.0" name="search tolerance minus value"/>

Must exist following the semantic mapping rules for the cvs.

**Comment by timosachsenberg** [comment]
so I guess we are currently exporting Da as tolerance unit in mzIdentML?
What do you suggest how we should proceed? I would actually not like to introduce any hacks/heuristics at this point but then we will get wrong mzIdentML if we convert from idXML to mzIdentML.

**Comment by mwalzer** [comment]
Well, as this is a commonly mandatory parameter, or at least set per default and there is already one  member in SearchParameter we could upgrade the member to a not-yet-full-fledged but better than now [CVTerm](ftp://ftp.mi.fu-berlin.de/pub/OpenMS/documentation/html/classOpenMS_1_1CVTerm.html) and chip in some accessor functions to mask the maybe little bit clunky handling of such CVTerm. If I have some spare time I will sketch some code.

**Comment by timosachsenberg** [comment]
CVTerm would be a flexible solution that also has disadvantages for internal data structures. If we allow the user to set arbitrary CV terms (e.g. also non-score terms) this is most certainly worse than just keeping the double and bool. On the other hand export and import is more straight forward with CVTerms. I personally prefer to go with the explicit modeling double + bool here but think it doesn't make a huge difference if we go for the CV term (but it would imply some extensive checks against the ontology and more code).

**Comment by mwalzer** [comment]
whats the status here?

**Comment by timosachsenberg** [comment]
I would include it as is maybe you could afterwards adapt the mzIdentML writer

**Comment by timosachsenberg** [comment]
travis passed except on mac for an unrelated clang warning

**Comment by mwalzer** [comment]
okay, I think we should merge that.

**Pull request merged**